conftest.py

```diff
@@ -43,30 +53,43 @@ def test_several_things(self):
     tests, CiTestCase's allowed_subp does take precedence (and we have
     TestDisableSubpUsageInTestSubclass to confirm that).
     """
-    should_disable = getattr(request, "param", True)
-    if should_disable:
-        if not isinstance(should_disable, (list, str)):
-            def side_effect(args, *other_args, **kwargs):
-                raise AssertionError("Unexpectedly used subp.subp")
-        else:
-            # Look this up before our patch is in place, so we have access to
-            # the real implementation in side_effect
-            real_subp = subp.subp
-
-            if isinstance(should_disable, str):
-                should_disable = [should_disable]
-
-            def side_effect(args, *other_args, **kwargs):
-                cmd = args[0]
-                if cmd not in should_disable:
-                    raise AssertionError(
-                        "Unexpectedly used subp.subp to call {} (allowed:"
-                        " {})".format(cmd, ",".join(should_disable))
-                    )
-                return real_subp(args, *other_args, **kwargs)
-
-        with mock.patch('cloudinit.subp.subp', autospec=True) as m_subp:
-            m_subp.side_effect = side_effect
-            yield
+    allow_subp_for = _closest_marker_args_or(request, "allow_subp_for", None)
+    # Because the mark doesn't take arguments, `allow_all_subp` will be set to
+    # [] if the marker is present, so explicit None checks are required
+    allow_all_subp = _closest_marker_args_or(request, "allow_all_subp", None)
+
+    if allow_all_subp is not None and allow_subp_for is None:
+        # Only allow_all_subp specified, don't mock subp.subp
+        yield
+        return
+
+    if allow_all_subp is None and allow_subp_for is None:
+        # No marks, default behaviour; disallow all subp.subp usage
+        def side_effect(args, *other_args, **kwargs):
+            raise AssertionError("Unexpectedly used subp.subp")
+
+    elif allow_subp_for is not None and allow_all_subp is not None:
```

Comment:
Can we keep the ordering of variable checks consistent as it makes me think harder when reviewing this
```suggestion
    elif allow_all_subp is not None and allow_subp_for is not None:
```

Reply:
Oh yeah, my bad! Fix pushed.